document-portal/document-portal-fuse.c

```diff
@@ -1897,7 +1899,25 @@ xdp_fuse_open (fuse_req_t req,
     return;
 
   path = fd_to_path (inode->physical->fd);
-  fd = open (path, open_flags, 0);
+  if (open_flags & O_NOFOLLOW)
+    {
+      int res;
+      char actual_path[PATH_MAX + 1];
+      /*
+       * `path` is a path to the fd entry in `/proc`, which is a symlink
+       * to the actual file. Opening it directly with `O_NOFOLLOW` will
+       * fail. So we should resolve it first then we can honour the no
+       * follow flag.
+       */
+      res = readlink (path, actual_path, sizeof (actual_path));
+      if (res == -1)
+        return xdp_reply_err (op, req, errno);
+      actual_path[res] = '\0';
+
+      fd = open (actual_path, open_flags, 0);
+    }
+  else
+    fd = open (path, open_flags, 0);
```

Comment:
I wonder if, instead of looking for the `open_flags` first, this function should do something like:

```c
  fd = open (path, open_flags, 0);
  if (errno == ELOOP && (open_flags & O_NOFOLLOW))
    {
      char actual_path[PATH_MAX + 1];
      ssize_t res;

      res = readlink (path, actual_path, sizeof (actual_path));

      ...

      fd = open (actual_path, open_flags, 0);
    }
```

The downside is that it would call `open()` twice.

Reply:
We know that here `path` points to the `/proc`, this is what `fd_to_path ()` returns. So we must resolve this.

Originally I was just resolving it unconditionally but then the tests started failing so I made the just to only make the exception for `O_NOFOLLOW` because we know this will break and kept the existing code path intact.